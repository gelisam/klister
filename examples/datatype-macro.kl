#lang "n-ary-app.kl"
(import "defun.kl")
(import (shift "n-ary-app.kl" 1))
(import (shift "lispy-do.kl" 1))
(import (shift "let.kl" 1))
(import (shift (only "list.kl" map) 1))


(datatype (List A)
  (nil)
  (:: A (List A)))

(defun snoc (xs x)
  (case xs
    [(nil) (:: x (nil))]
    [(:: y ys) (:: y (snoc ys x))]))

(defun reverse (xs)
  (case xs
    [(nil) (nil)]
    [(:: x xs) (snoc (reverse xs) x)]))

(define-macros
  ([list
     (lambda (stx)
       (syntax-case stx
         [(cons _ more)
          (syntax-case more
           [()
            (pure (replace-loc more '(nil)))]
           [(cons x xs)
            (pure (quasiquote/loc more (:: ,x ,(cons-list-syntax 'list xs xs))))])]))]))

(example (reverse (:: 1 (:: 2 (:: 3 (nil))))))

(example (reverse (list 1 2 3)))

(define-macros
  ([null (lambda (stx) (pure '(nil)))]))

(define-macros
  ([head
    (lambda (stx)
     (syntax-case stx
       [(list (_ x))
        (pure (quasiquote/loc stx (:: ,x xs)))]))]))

(example (case (reverse null) [null 'a]))

(example (case (reverse (list 1 2 3)) [(head x) x]))
         
