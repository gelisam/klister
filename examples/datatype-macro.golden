(:: 3 (:: 2 (:: 1 (nil)))) : (List Integer)
(:: 3 (:: 2 (:: 1 (nil)))) : (List Integer)
#[datatype-macro.kl:49.38-49.39]<a> : Syntax
3 : Integer
